### DISCUSSION TITLE
Terminal akışı

### DISCUSSION THREAD

**Discussion by BunyaminAkdeniz** [19/classroom#52] [Anonim] [answered]
Merhaba Hocam. Aklıma biraz anlamsız bi soru takıldı. Linux'ta kullandığım bazı toolslarda terminal ekranı aşağıya doğru kaymıyor da ekrana basılmış olan yazılar güncelleniyor. Örneğin yeni bir tools yüklenirken yükleme barı her artışta yeniden çıkmıyor. 
Bende bunun gibi ekrana sürekli eklenen değil de güncellenmesini istediğim, sarkacın yatay hareketini inceleyen örnek bir kod yazdım. Bu örnekteki gibi listenin her değişimde ekrana listeyi basmasını değil de listeyi terminal üzerinde sürekli güncellemesini ya da en azından diğer toolsların bunu nasıl yaptığını anlamak istiyorum. Umarım anlatabilmişimdir. 

```ruby
class Point
    def initialize x, speed
        @x=x
        @speed = speed
    end

    def speed_incrade speed
        return speed + 1
    end

    def speed_dec speed
        return speed - 1
    end

    def every_so_many_seconds(seconds)
        last_tick = Time.now
        loop do
          sleep 0.05
            if Time.now - last_tick >= seconds
              last_tick += seconds
                yield
            end
        end
    end

    
    def tiktak 
        a = Array.new(57){0}
        limit = 7
        every_so_many_seconds(0.01) do 
           a[@x] = 0 
            if @speed < limit
                @speed  = speed_incrade @speed
                @x += @speed 
                limit = 7
            else
                @x += @speed
                @speed  = speed_dec @speed
                limit = -7
                
            end 
                a[@x.to_i] = "XXXX"
                p a
        end
    end
end

x = Point.new 0,0
x.tiktak
```

**Comment by roktas** [comment, marked as answer]
UNIX terminali basittir ve standart olarak sona eklemeli (append) davranış gösterir.  Yani bir GUI widget'ı, örneğin grafik ortamda çizim yaptığın bir tuval (canvas) widget'ı gibi davranmaz.  Bu nedenle TUI (GUI'den mülhem: Text User Interface) olanakları sınırlıdır, alt seviye terminal çağrılarıyla kendin bir şeyler kurgulaman veya daha güzeli özel kitaplıklar kullanman gerekir.  "Spinner" dediğimiz widget'lar buna güzel  bir örnek.  Ruby özelinde şunu inceleyebilirsin: https://github.com/piotrmurach/tty-spinner  Bu kitaplık mekanizmayı anlaman için yararlı.  Amacın salt bir tür spinner değil de TTY üzerinde hakimiyet kurarak daha farklı şeyler yapmak ise aynı yazarın TTY kitaplıklarını inceleyebilirsin: https://ttytoolkit.org/

**Reply by BunyaminAkdeniz** [reply]
Teşekkür ederim Hocam. Tam aradığım şey
